File: src/garry_config.c
/*
 * Copyright (c) 2026 Matthew Hughes (Argylelabcoat)
 * Licensed under the BSD-3-Clause License.
 * See LICENSE file for details.
 */

/**
 * @file garry_config.c
 * @brief Default configuration factory for the storage engine.
 *
 * Bridges engine_settings (internal) to garry_config (public API),
 * providing callers with a sensible starting configuration.
 */

#include "garry/database.h"
#include "engine_settings.h"

/**
 * @brief Return a garry_config populated with default engine settings.
 *
 * Copies pool size, record size, page size, max transactions, and
 * max versions from the engine defaults into the public config struct.
 */
garry_config garry_config_default(void)
{
    garry_config c;
    garry_engine_settings s = garry_default_engine_settings();
    c.pool_size       = s.pool_size;
    c.max_record_size = s.max_record_size;
    c.page_size       = s.page_size;
    c.max_txns        = s.max_txns;
    c.max_versions    = s.max_versions;
    c.compression     = s.compression;
    c.btree_flags     = s.btree_flags;
    return c;
}

File: wrappers/c-sharp/GarrySharp/GarryConfig.cs
using System;
using System.Runtime.InteropServices;

namespace GarrySharp;

/// <summary>
/// Runtime configuration for a Garry database.
/// </summary>
public sealed class GarryConfig
{
    public int PoolSize { get; set; } = 8;
    public int MaxRecordSize { get; set; } = 16384;
    public int PageSize { get; set; } = 4096;
    public int MaxTxns { get; set; } = 4;
    public int MaxVersions { get; set; } = 64;
    public int MaxKeySize { get; set; } = 256;
    public int MaxSubscripts { get; set; } = 16;
    public bool UseCompression { get; set; } = false;
    public int BTreeFlags { get; set; } = 0;

    /// <summary>
    /// Convert to the native garry_config struct for P/Invoke.
    /// </summary>
    internal GarryConfigNative ToNative()
    {
        return new GarryConfigNative
        {
            pool_size = PoolSize,
            max_record_size = MaxRecordSize,
            page_size = PageSize,
            max_txns = MaxTxns,
            max_versions = MaxVersions,
            max_key_size = MaxKeySize,
            max_subscripts = MaxSubscripts,
            compression = UseCompression ? 1 : 0,
            btree_flags = BTreeFlags
        };
    }
}

/// <summary>
/// Native struct matching the C garry_config layout.
/// </summary>
[StructLayout(LayoutKind.Sequential)]
internal struct GarryConfigNative
{
    public int pool_size;
    public int max_record_size;
    public int page_size;
    public int max_txns;
    public int max_versions;
    public int max_key_size;
    public int max_subscripts;
    public int compression;
    public int btree_flags;
}
